### ISSUE TITLE
Support for Bukkit#createCommandSender

### ISSUE THREAD

**Issue by learliet** [JorelAli/CommandAPI#488]
### CommandAPI version

9.1.0

### Minecraft version

1.20.1

### Are you shading the CommandAPI?

Yes

### What I did

I encountered an issue when trying to use `Bukkit#createCommandSender` in combination with `Bukkit#dispatchCommand`. 
```kotlin
val myCommandSender = Bukkit.createCommandSender { }
Bukkit.dispatchCommand(myCommandSender, "<anycommandapicommand>")
```

### What actually happened

Executing commands from custom senders generated by `Bukkit#createCommandSender` results in a `NullPointerException` (see below) due to incorrect handling by the CommandAPI.

### What should have happened

The CommandAPI should support the special `CommandSender` returned by `Bukkit#createCommandSender`. 

### Server logs and CommandAPI config

```
Cannot invoke "[...].dev.jorel.commandapi.commandsenders.AbstractCommandSender.getSource()" 
because "this.val$sender" is null
```

### Other

Given [this method](https://github.com/JorelAli/CommandAPI/blob/master/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L379-L393), the CommandAPI assumes that `CommandSender` will only be the types provided by Spigot. However, on Paper, `Bukkit#createCommandSender` returns [`io.papermc.paper.commands.FeedbackForwardingSender`](https://github.com/PaperMC/Paper/blob/33bec7f20a5d3e20d9a8c0a16a1e5276add5d4fb/patches/server/0772-API-for-creating-command-sender-which-forwards-feedb.patch#L7), which extends `CommandSender`.

**Comment by JorelAli** [comment]
Is this just going to be a case of adding support for `io.papermc.paper.commands.FeedbackForwardingSender`, or should the CommandAPI export generic `CommandSender` classes?

**Comment by learliet** [comment]
In my personal case, adding support for `io.papermc.paper.commands.FeedbackForwardingSender` would be just fine. It would be greatly appreciated to implement this support first, regardless of whether generic `CommandSender` implementations are added later just to get support for custom command senders in place as quickly as possible.

As for generic `CommandSender` support, I’m not sure whether it is necessary or even appropriate. Then again, being able to register custom `CommandSender` classes seems like a useful niche feature at first glance.

**Comment by willkroboth** [comment]
I don't think supporting generic `CommandSender` implementations would be useful. Implementing the `CommandSender` class is not supported by Spigot ([SPIGOT-7023](https://hub.spigotmc.org/jira/browse/SPIGOT-7023)). You can ignore Spigot and create your own command sender anyway. In my experience, custom command senders do seem to work within the Bukkit CommandMap system. However, the bridge from Bukkit to Brigadier is incompatible with custom command senders. This is related to #477.

[`VanillaCommandWrapper#getListener`](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java?until=40945171bf8a5a234ed084386ef6b89070d2a4e3#66-87) is responsible for turning the Bukkit `CommandSender` into a Vanilla `CommandListenerWrapper`. This is the code that does that:

```java
public static CommandListenerWrapper getListener(CommandSender sender) {
    if (sender instanceof Entity) {
        if (sender instanceof CommandMinecart) {
            return ((EntityMinecartCommandBlock) ((CraftMinecartCommand) sender).getHandle()).getCommandBlock().createCommandSourceStack();
        }

        return ((CraftEntity) sender).getHandle().createCommandSourceStack();
    }
    if (sender instanceof BlockCommandSender) {
        return ((CraftBlockCommandSender) sender).getWrapper();
    }
    if (sender instanceof RemoteConsoleCommandSender) {
        return ((CraftRemoteConsoleCommandSender) sender).getListener().createCommandSourceStack();
    }
    if (sender instanceof ConsoleCommandSender) {
        return ((CraftServer) sender.getServer()).getServer().createCommandSourceStack();
    }
    if (sender instanceof ProxiedCommandSender) {
        return ((ProxiedNativeCommandSender) sender).getHandle();
    }

    throw new IllegalArgumentException("Cannot make " + sender + " a vanilla command listener");
}
```

This method only succeeds if the `CommandSender` is a block, console, entity, or proxy. These checks are actually much stricter, as the sender is usually cast to a CraftBukkit class, throwing a `ClassCastException` if the sender simply implements the Bukkit-API interfaces. You *can* still create custom command senders if you work with these restrictions, but [I don't recommend that](https://github.com/willkroboth/ConfigCommands/blob/fixOpConsole/ConfigCommands-nms/ConfigCommands-1_20/src/main/java/me/willkroboth/configcommands/nms/v1_20/opsenders/ConsoleOpSender1_20.java). Regardless, in order for a `CommandSender` to run Vanilla commands, it has to be one of these specific classes. So, it should be safe for the CommandAPI to assume the `CommandSender` objects passed into [`CommandAPIBukkit#wrapCommandSender`](https://github.com/JorelAli/CommandAPI/blob/master/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L379-L393) will only be those that are already being checked for.

This bug only occurs because Paper [added special logic](https://github.com/PaperMC/Paper/blob/b8edb0e1306b570f76d0e2b4f59cceee24b838fd/patches/server/0772-API-for-creating-command-sender-which-forwards-feedb.patch#L163-L178) to make their `FeedbackForwardingSender` work with `VanillaCommandWrapper#getListener`:

```diff
diff --git a/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java b/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
index 45e85f252acd72800956cc2a120564256d2de797..c3742b6c4abea173b38307048091fc56bd5051d7 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/VanillaCommandWrapper.java
@@ -83,6 +83,11 @@ public final class VanillaCommandWrapper extends BukkitCommand {
         if (sender instanceof ProxiedCommandSender) {
             return ((ProxiedNativeCommandSender) sender).getHandle();
         }
+        // Paper start
+        if (sender instanceof io.papermc.paper.commands.FeedbackForwardingSender feedback) {
+            return feedback.asVanilla();
+        }
+        // Paper end
 
         throw new IllegalArgumentException("Cannot make " + sender + " a vanilla command listener");
     }
```

So yeah, this should just be a case of adding support for `FeedbackForwardingSender` when running on Paper. Since the CommandAPI generates Vanilla commands, any other custom `CommandSender`, while technically possible, can't run CommandAPI commands anyway.

**Comment by JorelAli** [comment]
@willkroboth Thanks for the insight. Let's go ahead with adding support for `FeedbackForwardingSender` when running on Paper. Let's see if we can push this out for our next release (9.2.0).

**Comment by JorelAli** [comment]
Implemented for 9.2.0's release. I haven't been able to write an automatic test for this, however was able to verify it works with a manual test:

- Compile a plugin with these commands:

  ```java
  new CommandAPICommand("sayhi")
      .executes((sender, args) -> {
          Bukkit.broadcastMessage("hi");
      })
      .register();
  
  new CommandAPICommand("mycmd")
      .executes((sender, args) -> {
          CommandSender myCommandSender = Bukkit.createCommandSender(a -> {});
          Bukkit.dispatchCommand(myCommandSender, "sayhi");
      })
      .register();
  ```

- Run `sayhi` from the console (verify this prints `hi` in the console)
- Run `mycmd` from the console (verify that this now no longer crashes)

**Comment by JorelAli** [comment]
Implemented in 9.2.0.
